File: src/python/ordenamiento/dual_pivot_quicksort.py
"""
Dual-Pivot QuickSort - Implementacion en Python.
Fuente: GeeksforGeeks

Algoritmo QuickSort optimizado que utiliza dos pivotes.
Usado en Java 7+ Arrays.sort() para objetos.

Complejidad: O(n log n) promedio, O(n^2) peor caso
"""

from typing import List


def dual_pivot_quicksort(arr: List[int]) -> List[int]:
    """
    Ordena un arreglo usando el algoritmo Dual-Pivot QuickSort.

    Args:
        arr: Lista de enteros a ordenar

    Returns:
        Nueva lista ordenada en orden ascendente
    """
    if not arr:
        return []

    result = arr.copy()

    if len(result) <= 1:
        return result

    _dual_pivot_quicksort_recursive(result, 0, len(result) - 1)
    return result


def _dual_pivot_quicksort_recursive(arr: List[int], left: int, right: int) -> None:
    """Funcion recursiva auxiliar."""
    if left < right:
        pivot1, pivot2 = _partition(arr, left, right)

        _dual_pivot_quicksort_recursive(arr, left, pivot1 - 1)
        _dual_pivot_quicksort_recursive(arr, pivot1 + 1, pivot2 - 1)
        _dual_pivot_quicksort_recursive(arr, pivot2 + 1, right)


def _partition(arr: List[int], left: int, right: int) -> tuple:
    """Realiza la particion con dos pivotes."""
    if arr[left] > arr[right]:
        arr[left], arr[right] = arr[right], arr[left]

    pivot1 = arr[left]
    pivot2 = arr[right]

    i = left + 1
    lt = left + 1
    gt = right - 1

    while i <= gt:
        if arr[i] < pivot1:
            arr[i], arr[lt] = arr[lt], arr[i]
            lt += 1
            i += 1
        elif arr[i] > pivot2:
            arr[i], arr[gt] = arr[gt], arr[i]
            gt -= 1
        else:
            i += 1

    lt -= 1
    gt += 1
    arr[left], arr[lt] = arr[lt], arr[left]
    arr[right], arr[gt] = arr[gt], arr[right]

    return lt, gt

File: src/java/org/algoritmos/ordenamiento/DualPivotQuickSort.java
package org.algoritmos.ordenamiento;

/**
 * Dual-Pivot QuickSort
 *
 * Variante de QuickSort que usa dos pivotes para dividir en tres partes.
 *
 * Fuente: GeeksforGeeks
 *
 * Complejidad: O(n log n) promedio
 */
public class DualPivotQuickSort {

    /**
     * Ordena el arreglo usando Dual-Pivot QuickSort.
     *
     * @param array Arreglo a ordenar (se modifica in-place)
     * @return El mismo arreglo ordenado
     */
    public static int[] sort(int[] array) {
        if (array == null || array.length < 2) {
            return array;
        }

        sort(array, 0, array.length - 1);
        return array;
    }

    private static void sort(int[] array, int left, int right) {
        if (left < right) {
            if (array[left] > array[right]) {
                swap(array, left, right);
            }

            int pivot1 = array[left];
            int pivot2 = array[right];

            int i = left + 1;
            int k = right - 1;
            int j = i;

            while (j <= k) {
                if (array[j] < pivot1) {
                    swap(array, j, i);
                    i++;
                } else if (array[j] >= pivot2) {
                    while (array[k] > pivot2 && j < k) {
                        k--;
                    }
                    swap(array, j, k);
                    k--;
                    if (array[j] < pivot1) {
                        swap(array, j, i);
                        i++;
                    }
                }
                j++;
            }

            i--;
            k++;
            swap(array, left, i);
            swap(array, right, k);

            sort(array, left, i - 1);
            sort(array, i + 1, k - 1);
            sort(array, k + 1, right);
        }
    }

    private static void swap(int[] array, int i, int j) {
        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }
}
